Fix missing workspace cleanups after executing tests in oe.core.tests.resources

HeikoKlare opened the pull request:
Several test cases/classes in `org.eclipse.core.tests.resources` do not perform a proper cleanup after their execution. In consequence, the workspace is not in a clean state afterwards, e.g., in terms of remaining projects/files, changed workspace settings, and still registered or missing listeners.
This can lead to failures of subsequent test executions that assume a clean workspace state, depending on the execution order of tests. This particularly occurs when executing a bunch of test classes within Eclipse in a single run, but may also affect the Maven build.

Since tests should be isolated to be executable in any order without affecting other tests, I propose the following changes to cleanup several places I found when investigating test failures:
* Add missing removal of workspace listeners 
* Properly reset workspace properties such as the `autobuild` property in `teardown`
* Add missing calls to `teardown` because of unnecessarily overwritten methods, missing calls to the super method or missing extensions of an according test class ensuring cleanup (i.e. `ResourceTest`)
* Add removal of all files places in the workspace directory but not contained in projects

This missing isolation of test executions may also be a reason for the randomly occuring `Verifier has not yet been given a resource delta` test failures, such as #99 and #319. At least during local execution I had random occurrences of that failure because of interfering workspace listeners still registered from previously executed tests.

HeikoKlare commented on resources/tests/org.eclipse.core.tests.resources/src/org/eclipse/core/tests/internal/resources/WorkspacePreferencesTest.java:
```diff
@@ -16,13 +16,16 @@
+import org.eclipse.core.tests.session.WorkspaceSessionTestSuite;
 
-public class WorkspacePreferencesTest extends ResourceTest {
+public class WorkspacePreferencesTest extends WorkspaceSessionTest {
```
This should be a `WorkspaceSessionTest` to have the test run with a dedicated workspace setup. Otherwise, the import/export test for preferences leads to a complete reinitialization of the preferences, which miss some of the previously attached listeners. Subsequent tests can fail due to those missing listeners. Having a dedicated workspace setup seems to be the option for test isolation from my point of view.

jukzi commented:
Hi Haiko, for eclipse we typically squash our changes before review, please squash to a single commit to keep history small or create separate pull requests when changes are totally unrelated. 
The setAutobuild() may cause unwanted side affects when it happens during an autobuild. Enabling and disabling it may cause a new autobuild run. So should consider to wait for autobuild to finish (see org.eclipse.core.internal.events.BuildManager.waitForAutoBuild() / waitForAutoBuildOff).

HeikoKlare commented:
@jukzi Thank you for your instant feedback!
I have squashed the commits and updated the pull request with the single resulting commit. I also improved the restoration of the autobuild setting by properly waiting for the autobuild job to finish.

jukzi commented:
Please solve all failed tests. If any fail is unrelated to the change make sure there is a bug open for the random(?) failed test.

HeikoKlare commented:
Both failures are unrelated to these changes. An issue for one of them was already open (https://github.com/eclipse-platform/eclipse.platform/issues/103). I created another one for the second failure, which also occurs in unrelated pull requests (https://github.com/eclipse-platform/eclipse.platform/issues/373).

jukzi commented on resources/tests/org.eclipse.core.tests.resources/src/org/eclipse/core/tests/resources/ResourceTest.java:
```diff
@@ -406,8 +407,38 @@ protected void cleanup() throws CoreException {
+		deleteFilesInWorkspace();
+	}
+
+	private void deleteFilesInWorkspace() {
```
this deletes files in the Filesystem - not in the workspace. It would be better to remove them with workspace API. Otherwise the workspace will be out of sync with filesystem. Probably there is already helpercode for that somewher.e (isn't that line 402??!)

HeikoKlare replied:
Yes, the code deletes files in the filesystem and not in the workspace. This was my intention, which was obviously not properly reflected by the naming. The reason for that proposal is improper cleanup by different test classes, which lead to the Eclipse workspace being cleaned (by deleting the root), but files still remaining in the file system at the workspace location. In subsequent tests, this could lead to failures due to unexpected files being present.
My proposal was kind of a fallback solution for tests that somehow only clean the workspace but not all files in the workspace directory. However, I agree with you that this may not be the best option. I will figure out _why_ files are still present. One reason I saw is that tests delete the workspace root without forcing files in closed projects to be deleted, but there may also be other cases.

However, I will separate that from this pull request and make a distinct issue/pull request for the file cleanup topic. I have removed the according changes (i.e. the `deleteFilesInWorkspace` method in `ResourceTest`) from the pull request.

vogella commented:
@jukzi can this be merged? Looks like it improves the test setup and does not introduce new errors.

jukzi commented:
ok, looks like random fails only.
@HeikoKlare please keep an eye on https://download.eclipse.org/eclipse/downloads/index.html in the next days o see if it causes additional errors.

jukzi commented:
@HeikoKlare 
we now have TestBug294854.testDelete_02 failing on all platforms. Since i have seen this first time with this commit i ask you to investigate.
https://download.eclipse.org/eclipse/downloads/drops4/I20230329-1800/testResults.php
